### ISSUE TITLE
FiniteSets can be filled with other objects than numbers, but this is not supported by other classes.

### ISSUE THREAD

**Issue by warpfel** [sympy/sympy#9488]
```
>>> FiniteSet([Matrix([1,2,0]), Matrix([1,3,0])])
{Matrix([
[1],
[2],
[0]]), Matrix([
[1],
[3],
[0]])}
```

But

```
>>> L= FiniteSet([Matrix([1,2,0]), Matrix([1,3,0])])
>>> GramSchmidt(L)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python2.7/dist-packages/sympy/matrices/dense.py", line 1531, in GramSchmidt
    tmp = vlist[i]
TypeError: 'FiniteSet' object does not support indexing
>>>
```

In my opinion a set is a storage for all mathematical objects, and therefore it should also fulfill this job in sympy, or am I wrong ?

So (in my opinion), there should be added an exception to the FiniteSet class, so that it only can be filled with numbers for the moment, and, as an other issue, the set class should be improved, so that it could contain all kind of mathematical objects.

**Comment by certik** [comment]
I agree about adding the exception, since it's clearly not supported yet in other classes.

**Comment by warpfel** [comment]
Hi,

in the tests_sets the following line can be found:
´´´
s = FiniteSet((1, 2), Float, A, -5, x, 'eggs', x**2, Interval)
´´´
so this should be fixed, too, and should be replaced by something like

´´´
FiniteSet(3,2,complex('5+j'), Number(3), 3.47342, Symbol('a'), Matrix([3,4]), 'eggs')
raises(TypeError)
´´´

or am I wrong?

**Comment by amitsaha** [comment]
I don't think the issue here is that of objects other than numbers:

```
>>> from sympy import FiniteSet, GramSchmidt
>>> s = FiniteSet(1, 2, 3)
>>> s
{1, 2, 3}
>>> GramSchmidt(s)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.4/site-packages/sympy/matrices/dense.py", line 1483, in GramSchmidt
    tmp = vlist[i]
TypeError: 'FiniteSet' object does not support indexing

```

**Comment by warpfel** [comment]
Ok,

then the Problem lies in GramSchmidt. Sorry, I should have seen that. 

In my opinion its very natural, to  apply GramSchmidt on a Set, since Bases are sets. 
However, iterating trough a set, by mapping elements of a set to definite integer numbers, not really makes sense (since there is no canonic isomorphism :D). 

Therefore i would suggest, that we should (if you share my opinion, that GramSchmidt should work with sets, too) improve the GramSchmidt algorithm by iterating in the sense "for i in FiniteSet(...)", since this works with sets:

```
>>> for i in FiniteSet(Matrix([3,4,5]), Matrix([0,8,15])):
...     print(i)
... 
Matrix([
[ 0],
[ 8],
[15]])
Matrix([
[3],
[4],
[5]])
```

**Comment by leosartaj** [comment]
For the case of ``FiniteSets``, another possible solution can be implementing __getitem__. I tried it and I seem to get an answer.

**Comment by leosartaj** [comment]
Oops, but args are not ordered. That could be a problem. I don't really have much knowledge about ``GramSchmidt``.

**Comment by jksuom** [comment]
The documentation of ``GramSchmidt`` is a bit confusing.
```
def GramSchmidt(vlist, orthog=False):
    """
    Apply the Gram-Schmidt process to a set of vectors.
```
The docstring speaks of a set of vectors, but the argument is called ``vlist`` and a list is also returned.
Thus one is expected to call with a list argument (``GramSchmidt(list(L))``).

I think this is a better solution than changing ``GramSchmidt`` to work with sets, since the Gram-Schmidt process relies on the order of vectors in the sequence. Otherwise the result would be  unpredictable and possibly implementation dependent.

**Comment by oscarbenjamin** [comment]
The expression in the original post here raises now. I think that's because the signature of FiniteSet has changed so:
```
In [13]: L = FiniteSet(Matrix([1,2,0]), Matrix([1,3,0]))                                                 

In [14]: L                                                                                               
Out[14]: 
⎧⎡1⎤  ⎡1⎤⎫
⎪⎢ ⎥  ⎢ ⎥⎪
⎨⎢2⎥, ⎢3⎥⎬
⎪⎢ ⎥  ⎢ ⎥⎪
⎩⎣0⎦  ⎣0⎦⎭

In [15]: GramSchmidt(L)                                                                                  
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-15-e1ef40d009e6> in <module>
----> 1 GramSchmidt(L)

~/current/sympy/sympy.git/sympy/matrices/dense.py in GramSchmidt(vlist, orthonormal)
   1212     m = len(vlist)
   1213     for i in range(m):
-> 1214         tmp = vlist[i]
   1215         for j in range(i):
   1216             tmp -= vlist[i].project(out[j])

TypeError: 'FiniteSet' object does not support indexing

In [16]: GramSchmidt(list(L))                                                                            
Out[16]: 
⎡⎡1⎤  ⎡-2/5⎤⎤
⎢⎢ ⎥  ⎢    ⎥⎥
⎢⎢2⎥, ⎢1/5 ⎥⎥
⎢⎢ ⎥  ⎢    ⎥⎥
⎣⎣0⎦  ⎣ 0  ⎦⎦
```
Everything is fine then except the docstring for Gram-Schmidt which should be updated to say that the input vectors should be in a sequence and that it returns a list (rather than unordered sets).

**Comment by oscarbenjamin** [comment]
Marking as easy to fix

**Comment by Smit-create** [comment]
On the current master I get:
```
>>> from sympy import *
>>> L = FiniteSet(Matrix([1,2,0]), Matrix([1,3,0]))
>>> L
FiniteSet(Matrix([
[1],
[2],
[0]]), Matrix([
[1],
[3],
[0]]))
>>> GramSchmidt(L)
[Matrix([
[1],
[2],
[0]]), Matrix([
[-2/5],
[ 1/5],
[   0]])]
```
Should this closed with a test case? @oscarbenjamin

**Comment by oscarbenjamin** [comment]
This was fixed in #16456. I test could be added I guess.

**Comment by sidhu1012** [comment]
@oscarbenjamin I have added a test case for this issue.
